crates/kubelet/src/pod.rs

```diff
@@ -85,27 +86,88 @@ impl Pod {
         Some(self.annotations().get(key)?.as_str())
     }
 
-    /// Patch the pod status to update the phase.
-    pub async fn patch_status(&self, client: APIClient, phase: &Phase) {
-        let status = serde_json::json!(
+    /// Patch the pod status using the given status information.
+    pub async fn patch_status(&self, client: APIClient, status: Status) {
+        let name = self.name();
+        let api: Api<KubePod> = Api::namespaced(client, self.namespace());
+        let current_status = match api.get(name).await {
+            Ok(p) => match p.status {
+                Some(s) => s,
+                None => {
+                    error!("Pod is missing status information. This should not occur");
+                    return;
+                }
+            },
+            Err(e) => {
+                error!("Unable to fetch current status of pod {}, aborting status patch (will be retried on next status update): {:?}", name, e);
+                return;
+            }
+        };
+
+        // This section figures out what the current phase of the pod should be
+        // based on the container statuses
+        let current_statuses = status
+            .container_statuses
+            .into_iter()
+            .map(|s| s.to_kubernetes(name.to_owned()))
+            .collect::<Vec<KubeContainerStatus>>();
+        // Filter out any ones we are updating and then combine them all together
+        let mut container_statuses = current_status
+            .container_statuses
+            .unwrap_or_default()
+            .into_iter()
+            .filter(|s| {
+                current_statuses
+                    .iter()
+                    .find(|&other| other.name == s.name)
+                    .is_none()
+            })
+            .collect::<Vec<KubeContainerStatus>>();
+        container_statuses.extend(current_statuses);
+        let mut num_terminated: usize = 0;
```

Comment:
`terminated` can have a connotation of failure. Is `num_succeed` a better name?

Reply:
Yes it is. Will change